Update dynamic.asciidoc (#1504)
vaadin/docs: articles/routing/dynamic.asciidoc

= Registering Routes Dynamically

> In addition to registering routes and route templates using the <<index#, `@Route`>> annotation, you can add and remove routes dynamically at runtime. 

Reviewer: ```suggestion
In addition to registering routes and route templates using the <<index#, `@Route`>> annotation, you can add and remove routes dynamically at runtime.
```

* all users using the *application* scope, or
* only certain active users using the *session* scope.

You can access the scope using the static methods [methodname]`forSessionScope()` and [methodname]`forApplicationScope()` from the [classname]`RouteConfiguration` class.

[NOTE]
All components annotated with `@Route` are added to the application scope at startup, unless the <<Dynamic Registration of @Route Annotated Classes, `registerAtStartup`>> element has been set to `false`.

== Configuring User-Specific Routes

You can add and remove routes for certain users; for example, based on their access rights.

*Example*: Adding *admin* and *home* views for users with an active session.

[source,java]
----
RouteConfiguration.forSessionScope()
        .setRoute(
                "admin", //path
                AdminView.class //navigation target
         );

// parent layouts can be given as a vargargs parameter
RouteConfiguration.forSessionScope()
        .setRoute(
                "home", //path
                HomeView.class, //navigation target
                MainLayout.class //one or more parents
        );
----

> Any registered `@Route` in the application scope can be overridden for a specific user in the session scope.

Reviewer: I feel that this suggestion hides a couple of important terms in the original formulation. First, in the original sentence, it was clear that we are speaking about "routes with the same path". Also, the original formulation implies that an application-scoped route could be defined dynamically or through the `@Route` annotation (whereas in the reformulated suggestion, only the `@Route` way is mentioned).

Author: Hi Tarek, thanks for the comment.

The problem with the original text was that it felt like reading the same sentence repeated twice.

> First, in the original sentence, it was clear that we are speaking about "routes with the same path".

I felt that "_routes with the same path_" is redundant when the word "_override_" was already being used. Whenever I read "_override_", I would automatically assume that you must be overriding something that is the same from the parent or wider scope. Including "_with the same path_" is also correct, albeit a little redundant.

For example:
In the sentence, "_Please override the method `public void doSomething()` with the same method name_", the fragment "_with the same method name_" is redundant because the overriding method must already have the same name.

> Also, the original formulation implies that an application-scoped route could be defined dynamically or through the `@Route` annotation

The original sentence said "This means that any **statically** registered `@Route`", so it definitely didn't read like both `dynamic` **and** `static`. 

> whereas in the reformulated suggestion, only the `@Route` way is mentioned

The reformulated sentence actually omitted both `dynamic` and `static`, so it can cover both cases.

The routes in the session scope are accessible for the current user only for as
long as the session is valid.

> 
> When the session is invalidated by the user logging out, the session-scoped routes are no longer available.
> It is unnecessary to manually remove these routes.

Reviewer: ```suggestion
When the session is invalidated by the user logging out, the session-scoped routes are no longer available.
It is unnecessary to manually remove these routes.
```

configuration.removeRoute("admin");
----

[NOTE]
Removing a route in the session scope that had previously overridden a route in the application scope makes the application-scoped route accessible once again.

[NOTE]
When dynamically registering a route, any annotations on classes *are ignored*, except when the method used contains `Annotated`; for example, [methodname]`setAnnotatedRoute()`.

> See <<Dynamic Registration of @Route Annotated Classes>> below for more.

Reviewer 2: ```suggestion
See <<Dynamic Registration of @Route Annotated Classes>> later in this article for more.
```
